src/analysis/constant_speed.lean

```diff
@@ -0,0 +1,291 @@
+  rw has_constant_speed_on_with_iff_variation_on_from_to_eq at hfφ,
+  symmetry,
+  calc  (y - x) * l
+      = l * (y - x) : by rw mul_comm
+  ... = variation_on_from_to (f ∘ φ) s x y : (hfφ.2 xs ys).symm
+  ... = variation_on_from_to f (φ '' s) (φ x) (φ y) :
+    variation_on_from_to.comp_eq_of_monotone_on f φ φm xs ys
+  ... = l' * (φ y - φ x) : hf.2 ⟨x,xs,rfl⟩ ⟨y,ys,rfl⟩
+  ... = (φ y - φ x) * l' : by rw mul_comm,
+end
+
+/-- `f` has unit speed on `s` if it is linearly parameterized by `l = 1` on `s`. -/
+def has_unit_speed_on (f : ℝ → E) (s : set ℝ) := has_constant_speed_on_with f s 1
+
+lemma has_unit_speed_on.union {t : set ℝ} {x : ℝ}
+  (hfs : has_unit_speed_on f s) (hft : has_unit_speed_on f t)
+  (hs : is_greatest s x) (ht : is_least t x) : has_unit_speed_on f (s ∪ t) :=
+has_constant_speed_on_with.union hfs hft hs ht
+
+lemma has_unit_speed_on.Icc_Icc {x y z : ℝ}
+  (hfs : has_unit_speed_on f (Icc x y)) (hft : has_unit_speed_on f (Icc y z)) :
+  has_unit_speed_on f (Icc x z) :=
+has_constant_speed_on_with.Icc_Icc hfs hft
+
+/--
+If both `f` and `f ∘ φ` have unit speed (on `t` and `s` respectively) and `φ`
+monotonically maps `s` onto `t`, then `φ` is just a translation (on `s`).
+-/
+lemma unique_unit_speed {φ : ℝ → ℝ} (φm : monotone_on φ s)
+  (hfφ : has_unit_speed_on (f ∘ φ) s) (hf : has_unit_speed_on f (φ '' s))
+  ⦃x : ℝ⦄ (xs : x ∈ s) : eq_on φ (λ y, (y - x) + (φ x)) s :=
+begin
+  dsimp only [has_unit_speed_on] at hf hfφ,
+  convert has_constant_speed_on_with.ratio one_ne_zero φm hfφ hf xs,
+  simp only [nonneg.coe_one, div_self, ne.def, one_ne_zero, not_false_iff, one_mul],
+end
+
+/--
+If both `f` and `f ∘ φ` have unit speed (on `Icc 0 t` and `Icc 0 s` respectively)
+and `φ` monotonically maps `Icc 0 s` onto `Icc 0 t`, then `φ` is the identity on `Icc 0 s`
+-/
+lemma unique_unit_speed_on_Icc_zero {s t : ℝ} (hs : 0 ≤ s) (ht : 0 ≤ t)
+  {φ : ℝ → ℝ}  (φm : monotone_on φ $ Icc 0 s) (φst : φ '' (Icc 0 s) = (Icc 0 t))
+  (hfφ : has_unit_speed_on (f ∘ φ) (Icc 0 s))
+  (hf : has_unit_speed_on f (Icc 0 t)) : eq_on φ id (Icc 0 s) :=
+begin
+  rw ←φst at hf,
+  convert unique_unit_speed φm hfφ hf ⟨le_rfl, hs⟩,
+  have : φ 0 = 0, by
+  { obtain ⟨x,xs,hx⟩ :=  φst.rec_on (surj_on_image φ (Icc 0 s)) ⟨le_rfl, ht⟩,
+    exact le_antisymm (hx.rec_on (φm ⟨le_rfl,hs⟩ xs xs.1))
+                      (φst.rec_on (maps_to_image φ (Icc 0 s)) (⟨le_rfl, hs⟩)).1, },
+  simp only [tsub_zero, this, add_zero],
+  refl,
+end
+
+/--
+The natural parameterization of `f` on `s`, which, if `f` has locally bounded variation on `s`,
+* has unit speed on `s`
+  (by `natural_parameterization_has_unit_speed`).
+* composed with `variation_on_from_to f s a`, is at distance zero from `f`
+  (by `natural_parameterization_edist_zero`).
+-/
+noncomputable def natural_parameterization (f : α → E) (s : set α) (a : α) : ℝ → E :=
+f ∘ (@function.inv_fun_on _ _ ⟨a⟩ (variation_on_from_to f s a) s)
+
+lemma natural_parameterization_edist_zero {f : α → E} {s : set α}
```

Comment:
the lemma name should match the order of words in the declaration, in this lemma and the next one.